bsd/backends.py

```diff
@@ -0,0 +1,50 @@
+from bs4 import BeautifulSoup
+from django.contrib.auth.models import User
+from StringIO import StringIO
+from xml.etree.ElementTree import ElementTree
+from .api import Bsd
+
+
+# TODO: what is best way to do this?
+api = Bsd().api
+
+class BSDAuthenticationBackend:
+
+    def authenticate(self, username=None, password=None):
+
+        # Basic email validation
+        if not '@' in username:
+            return None
+
+        # Check credentials via BSD api
+        apiResult = api.account_checkCredentials(username, password)
+
+        # Parse and validate response
+        tree = ElementTree().parse( StringIO(apiResult.body) )
+        try:
+            cons = tree.find('cons')
+            # assertions copied from hydra app
+            assert cons is not None
+            assert cons.find('has_account').text == "1"
+            assert cons.find('is_banned').text == "0"
+
+            # Get user from db that matches email or create new one if needed
+            try:
+                user = User.objects.get(email=username)
+            except User.DoesNotExist:
+                # Create user but dont set password since we use BSD password
```

Comment:
Should we be creating a user here or is this just a holdover from Hydra? Shouldn't we say "user not found, create an account" or something along those lines? How are we creating a user without a password?

Reply:
This is not from Hydra. Since we need to return a User model for the authentication backend methods, this seemed like the simplest most standard approach to use. This way we have standard User objects and post-authentication logic can just handle BSD users in normal django fashion by default.

`If no password is provided, set_unusable_password() will be called.`
https://docs.djangoproject.com/en/1.10/ref/contrib/auth/#django.contrib.auth.models.UserManager.create_user
https://docs.djangoproject.com/en/1.10/ref/contrib/auth/#django.contrib.auth.models.User.set_unusable_password

Reply:
The reason we don't say "user not found, create an account" is because they have already passed BSD api credentials check, and we are just looking for a local user record in db at this point.

In other words, there is a decoupling of bsd authentication step and db user record storage. So the password check happens in bsd but we still have a user record in local db.